prow/plugins/verify-owners/verify-owners.go

```diff
@@ -0,0 +1,137 @@
+	return &pluginhelp.PluginHelp{
+			Description: "The owners plugin validates formatting and 'labels' sections of OWNERS files modified in a PR. On failure it automatically adds a 'do-not-merge/invalid-OWNERS' label to the PR.",
+		},
+		nil
+}
+
+type ownersClient interface {
+	LoadRepoOwners(org, repo, base string) (repoowners.RepoOwnerInterface, error)
+}
+
+type githubClient interface {
+	AddLabel(owner, repo string, number int, label string) error
+	CreateComment(owner, repo string, number int, comment string) error
+	GetPullRequestChanges(org, repo string, number int) ([]github.PullRequestChange, error)
+}
+
+func handlePullRequest(pc plugins.PluginClient, pre github.PullRequestEvent) error {
+	if pre.Action != github.PullRequestActionOpened && pre.Action != github.PullRequestActionReopened && pre.Action != github.PullRequestActionSynchronize {
+		return nil
+	}
+	return handle(pc.GitHubClient, pc.GitClient, pc.Logger, &pre, pc.PluginConfig.Owners.LabelsBlackList)
+}
+
+func handle(ghc githubClient, gc *git.Client, log *logrus.Entry, pre *github.PullRequestEvent, labelsBlackList []string) error {
+	org := pre.Repo.Owner.Login
+	repo := pre.Repo.Name
+	var errMsg error
+
+	// Get changes.
+	changes, err := ghc.GetPullRequestChanges(org, repo, pre.Number)
+	if err != nil {
+		return fmt.Errorf("error getting PR changes: %v", err)
+	}
+
+	// List modified OWNERS files.
+	var modifiedOwnersFiles []string
+	for _, change := range changes {
+		if filepath.Base(change.Filename) == ownersFileName {
+			modifiedOwnersFiles = append(modifiedOwnersFiles, change.Filename)
+		}
+	}
+	if len(modifiedOwnersFiles) == 0 {
+		return nil
+	}
+
+	// Clone the repo, checkout the PR.
+	r, err := gc.Clone(pre.Repo.FullName)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := r.Clean(); err != nil {
+			log.WithError(err).Error("Error cleaning up repo.")
+		}
+	}()
+	if err := r.CheckoutPullRequest(pre.Number); err != nil {
+		return err
+	}
+
+	// Check each OWNERS file.
+	for _, f := range modifiedOwnersFiles {
+		// Try to load OWNERS file.
+		simple, _, err := repoowners.ParseOwners(filepath.Join(r.Dir, f))
+		if err != nil {
+			errMsg = fmt.Errorf("error occured parsing the OWNERS file: %v", err)
+			break
+		}
+		// Check labels against blacklist
+		if sets.NewString(simple.Config.Labels...).HasAny(labelsBlackList...) {
+			errMsg = fmt.Errorf("OWNERS file contains blacklisted labels")
+			break
+		}
+		// Check approvers isn't empty
+		if filepath.Dir(f) == "." && len(simple.Config.Approvers) == 0 {
+			errMsg = fmt.Errorf("no approvers defined for the repo")
+			break
+		}
+	}
+	// React if we saw something.
+	if errMsg != nil {
+		log.Infof("Adding %s label.", invalidOwnersLabel)
+		if err := ghc.AddLabel(org, repo, pre.Number, invalidOwnersLabel); err != nil {
+			return err
+		}
+		resp := fmt.Sprintf("Your OWNERS file is invalid: %v", errMsg)
```

Comment:
We should track which file is invalid and leave a review comment on that file like the `lint` plugin does.

Reply:
Like that? I had to add CreateReview to fakegithub client for the tests to pass...